src/test/java/com/lginterview/integration/CourseApiTests.java

```diff
@@ -0,0 +1,217 @@
+        //arrange
+        String courseSaveRequest = "{\"userId\":\"1\",\"courseId\":\"23\",\"type\":\"save\"," +
+                "\"score\":\"45\",\"timestamp\" :\"2017-06-28 14:21:45.375193+03:00\"}";
+
+        //act
+        ResponseEntity<String> courseSaveResponse =
+                template.postForEntity(
+                        "/status/course",
+                        new HttpEntity<>(courseSaveRequest, headers),
+                        String.class);
+
+        //assert
+        assertNotNull(courseSaveResponse);
+        assertEquals(HttpStatus.NOT_FOUND, courseSaveResponse.getStatusCode());
+    }
+
+    @Test
+    public void Given_CourseFinishRequestForExistingCourse_When_requestCourseModification_Then_CourseIsFinished() {
+        //arrange
+        Course newCourse = new Course();
+        long newCourseId = courseRepository.save(newCourse).getId();
+
+        String courseSaveRequest = "{\"userId\":\"1\",\"courseId\":\"" + Long.toString(newCourseId) +
+                "\",\"type\":\"finish\"," +
+                "\"score\":\"45\",\"timestamp\" :\"2017-06-28T14:21:45.375193+03:00\"}";
+
+        //act
+        ResponseEntity<Long> courseFinishResponse =
+                template.postForEntity(
+                        "/status/course",
+                        new HttpEntity<>(courseSaveRequest, headers),
+                        Long.class);
+
+        //assert
+        assertNotNull(courseFinishResponse);
+        assertEquals(HttpStatus.OK, courseFinishResponse.getStatusCode());
+        assertEquals(newCourseId, courseFinishResponse.getBody().longValue());
+    }
+
+    @Test
+    public void Given_CourseFinishRequestForNotExistingCourse_When_requestCourseModification_Then_NotFoundIsReturned() {
+        //arrange
+        String courseSaveRequest = "{\"userId\":\"1\",\"courseId\":\"23\",\"type\":\"finish\"," +
+                "\"score\":\"45\",\"timestamp\" :\"2017-06-28 14:21:45.375193+03:00\"}";
+
+        //act
+        ResponseEntity<String> courseFinishResponse =
+                template.postForEntity(
+                        "/status/course",
+                        new HttpEntity<>(courseSaveRequest, headers),
+                        String.class);
+
+        //assert
+        assertNotNull(courseFinishResponse);
+        assertEquals(HttpStatus.NOT_FOUND, courseFinishResponse.getStatusCode());
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void Given_CourseStatusRequestForExistingCourse_When_requestCourseStatus_Then_StatusIsReturned() {
+        //arrange
+        List<Integer> scores = fixture.collections().createCollection(List.class, Integer.class);
+
+        User user = userRepository.findOne(1L);
+
+        Course newCourse = getNewCourse(scores, user);
+        long newCourseId = courseRepository.save(newCourse).getId();
+
+        int expectedBestScore = scores.stream().mapToInt(score -> score).max().orElse(0);
```

Comment:
[MS][Major] I don't like the idea of generating random data (scores in this case) and copying logic from production code. You should have a couple edge-cases (empty list, one element, big integers, big list etc.) to properly test it.

Example: if someone will break the code so that it can't deal with empty lists (because I guess 'scores' will always have at least one element) your test will not find this bug.

Reply:
I Agree